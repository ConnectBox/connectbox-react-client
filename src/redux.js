export function getContent (contentPath, callback) {
  return { type: 'CONTENT_FETCH_REQUESTED', payload: { contentPath, callback } }
}

export function sendMessage (message, callback) {
  return { type: 'MESSAGE_SEND_REQUESTED', message }
}

export function getMessages (callback) {
  return { type: 'MESSAGES_FETCH_REQUESTED' }
}

export function fetchNick (callback) {
  return { type: 'FETCH_NICK_REQUESTED' }
}

export function saveNick (nick, callback) {
  return { type: 'SAVE_NICK_REQUESTED', nick }
}

export function fetchTextDirection (callback) {
  return { type: 'FETCH_TEXT_DIRECTION_REQUESTED' }
}

export function saveTextDirection (textDirection, callback) {
  return { type: 'SAVE_TEXT_DIRECTION_REQUESTED', textDirection }
}

export function getNewMessages (callback) {
  return { type: 'NEW_MESSAGES_FETCH_REQUESTED' }
}

export function setConfigPath (configPath, callback) {
  let path = configPath
  if (!path) {
    path = '/config/default.json'
  }
  return { type: 'SET_CONFIG_PATH', payload: {configPath: path} }
}

export function clearMessageNotifications (callback) {
  return { type: 'CLEAR_MESSAGE_NOTIFICATIONS' }
}

export function toggleChatPanel (showing, callback) {
  return { type: 'TOGGLE_CHAT_PANEL', showing }
}

export const initialState = {
  error: undefined,
  contentPath: '',
  content: undefined
}

const fileExtension = (fname) => {
  return fname.substr((~-fname.lastIndexOf('.') >>> 0) + 2)
}

const parseIcoMetadata = (prefix, filename, dirMap) => {
  let ext = fileExtension(filename)

  const metadata = filename.substring(prefix.length)

  if (ext) {
    let dirName = metadata.substring(0, metadata.lastIndexOf('.'))
    if (dirMap[dirName]) {
      let dir = dirMap[dirName]
      dir.icon_file = filename
    }
  } else {
    let parts = metadata.split('_')
    let dirName = parts[0]
    let ico = parts[1]

    if (dirMap[dirName]) {
      let dir = dirMap[dirName]
      dir.icon_class = ico
    }
  }
}

const postProcessContent = (content, contentPath, state) => {
  const {config} = state
  const isTopLevel = contentPath === '' || contentPath === '/'
  const dirMap = {}
  content.map((f) => {
    f.isTopLevel = isTopLevel
    f.mtime = new Date(f.mtime)
    if (f.type === 'file') {
      f.ext = fileExtension(f.name)
      if (f.isTopLevel) {
        f.isTopLevel = false
      }
    } else {
      dirMap[f.name] = f
    }

    return f
  })

  const excludeFilters = config.Client.excluded_files ? config.Client.excluded_files.map((pattern) => {
    return RegExp(pattern, 'i')
  }) : []

  excludeFilters.push(/^\..*/)
  // Remove excluded files
  let files = content.filter((f) => {
    const excluded = excludeFilters.reduce((a, b) => a || RegExp(b).test(f.name), false)
    if (excluded) {
      return false
    }

    if (f.name.startsWith(config.Client.icon_prefix)) {
      parseIcoMetadata(config.Client.icon_prefix, f.name, dirMap)
      return false
    } else {
      return true
    }
  })

  let topLevelFiles = null
  if (isTopLevel) {
    topLevelFiles = files.filter(f => f.type === 'file')
    files = files.filter(f => f.type === 'directory')
  }

  return [files, topLevelFiles]
}

const postProcessStats = (stats) => {
  let popularFiles = []
  let max = 0
  let min = 2147483647
  popularFiles = ((stats && stats['year']) || []).map(function (f) {
    var resource = f.resource
    var parts = resource.split('/')
    f.name = decodeURIComponent(parts[parts.length - 1])
    f.ext = fileExtension(f.name)
    if (f.count > max) {
      max = f.count
    }
    if (f.count < min) {
      min = f.count
    }
    return f
  })

  var maxMin = (max - min)
  if (maxMin <= 0) {
    maxMin = 1
  }
  popularFiles = popularFiles.map(function (f) {
    f.score = Math.round(4 * (f.count - min) / maxMin + 1)
    return f
  })

  return popularFiles
}

export default function reducer (state = initialState, action) {
  const handler = handlers[action.type]
  return handler ? handler(state, action) : state
}

const handlers = {
  'CONTENT_FETCH_SUCCEEDED': (state, action) => {
    const { content, contentPath } = action
    const [files, topLevelFiles] = postProcessContent(content, contentPath, state)
    return { ...state, content: files, topLevelFiles, contentPath, loading: false }
  },

  'CONTENT_FETCH_FAILED': (state, action) => {
    return { ...state, error: action.message, loading: false }
  },

  'CONFIG_FETCH_SUCCEEDED': (state, action) => {
    return { ...state, config: action.config, needsConfig: false }
  },

  'CONFIG_FETCH_FAILED': (state, action) => {
    return { ...state, error: action.message }
  },

  'SET_CONFIG_PATH': (state, action) => {
    return { ...state, configPath: action.payload.configPath }
  },

  'LOADING_START': (state, action) => {
    return { ...state, loading: true }
  },

  'STATS_FETCH_SUCCEEDED': (state, action) => {
    const {stats} = action
    const popularFiles = postProcessStats(stats)
    return { ...state, popularFiles }
  },

  'STATS_FETCH_FAILED': (state, action) => {
    return { ...state, error: action.message }
  },

  'MESSAGES_FETCH_START': (state, action) => {
    return { ...state, error: action.message, loadingMessages: true }
  },

  'MESSAGES_FETCH_SUCCEEDED': (state, action) => {
    const { messages } = action
    const mention = action.checkMentions ? (state.mention ? true : messages.reduce((mentioned, message) => {
      if (!mentioned) {
        if (message.body.indexOf(`@${state.nick}`) !== -1) {
          return true
        }
        return false
      }
      return true
    }, false)) : false
    return {
      ...state,
      chatOffline: false,
      maxMessageId: messages.length > 0 ? messages.reduce((max, message) =>
        message.id > max ? message.id : max, 0) : state.maxMessageId,
      messages: Object.assign({}, state.messages, messages.reduce((map, message) => {
        map[message.id] = message
        return map
      }, {})),
      mention,
      loadingMessages: false,
      newMessages: !mention && state.maxMessageId !== undefined && (state.newMessages || (messages && messages.length > 0))
    }
  },

  'MESSAGES_FETCH_FAILED': (state, action) => {
    return { ...state, chatOffline: true, loadingMessages: false }
  },

  'MESSAGE_SEND_SUCCEEDED': (state, action) => {
    return {
      ...state,
      sentMessages: [
        ...state.sentMessages,
        action.messageId
      ]
    }
  },

  'MESSAGE_SEND_FAILED': (state, action) => {
    return { ...state, error: action.message }
  },

  'FETCH_NICK_SUCCEEDED': (state, action) => {
    return { ...state, nick: action.nick }
  },

  'SAVE_NICK_SUCCEEDED': (state, action) => {
    return { ...state, nick: action.nick }
  },

  'FETCH_TEXT_DIRECTION_SUCCEEDED': (state, action) => {
    return { ...state, textDirection: action.textDirection }
  },

  'SAVE_TEXT_DIRECTION_SUCCEEDED': (state, action) => {
    return { ...state, textDirection: action.textDirection }
  },

  'TOGGLE_CHAT_PANEL': (state, action) => {
    return { ...state, chatPanelShowing: action.showing }
  },

  'CLEAR_MESSAGE_NOTIFICATIONS': (state, action) => {
    return { ...state, mention: false, newMessages: false }
  }

}
