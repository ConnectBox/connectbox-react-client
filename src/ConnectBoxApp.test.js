/* global describe, it, expect, jest */
import React from 'react'
import ReactDOM from 'react-dom'
import App, {ConnectBoxApp} from './ConnectBoxApp'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import { MemoryRouter as Router } from 'react-router-dom'
import createSagaMiddleware from 'redux-saga'
import reducer from './redux'
import { createStore, applyMiddleware, compose } from 'redux'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

const config = {
  Content: {
    contentRoute: '/content'
  },
  Client: {
    tail_slash: true,
    display_root_folder_names: true,
    icon_prefix: '_icon_',
    stats_file: 'stats.top10.json'
  }
}
const iconMetadata = {ids: [], names: {}}
const contentPath = ''
const sagaMiddleware = createSagaMiddleware()

describe('ConnectBoxApp tests', () => {
  it('renders without crashing', () => {
    const makeMockStore = configureMockStore([sagaMiddleware])

    const mockStore = makeMockStore({
      content: [],
      error: undefined,
      config,
      configPath: '/config/default.json',
      contentPath,
      iconMetadata,
      needsConfig: false,
      loading: false,
      popularFiles: null
    }, reducer)

    const div = document.createElement('div')
    ReactDOM.render(<Provider store={mockStore}><Router><App /></Router></Provider>, div)
  })

  it('shallow render succeeds', () => {
    shallow(
      <ConnectBoxApp
        contentPath='/'
        config={config}
        content={[]}
        getContent={jest.fn()}
        setConfigPath={jest.fn()}
        history={{}}
        iconMetadata={{}}
        loading={false}
        location={{}} />)
  })

  it('renders content fully', () => {
    const mockDate = new Date('Thu, 29 Jun 2017 02:23:21 GMT')
    const prepopulatedState = { content:
    [ { name: 'Help',
      type: 'directory',
      mtime: mockDate,
      isTopLevel: true,
      icon_file: '_icon_Help.png' },
    { name: 'Music',
      type: 'directory',
      mtime: mockDate,
      isTopLevel: true },
    { name: 'Videos',
      type: 'directory',
      mtime: mockDate,
      isTopLevel: true,
      icon_class: 'fa-video-camera' },
    { name: 'standard-folder',
      type: 'directory',
      mtime: mockDate,
      isTopLevel: true },
    { name: 'throughput-test',
      type: 'directory',
      mtime: mockDate,
      isTopLevel: true },
    { name: 'written',
      type: 'directory',
      mtime: mockDate,
      isTopLevel: true,
      icon_class: 'fa-book' } ],
      error: undefined,
      config:
      { Content: { contentRoute: '/content' },
        Client:
        { tail_slash: true,
          display_root_folder_names: true,
          icon_prefix: '_icon_',
          stats_file: 'stats.top10.json' } },
      configPath: '/config/default.json',
      contentPath: '',
      iconMetadata: { ids: [], names: {} },
      needsConfig: false,
      loading: false,
      popularFiles:
      [ { resource: '/content/Music/Vivaldi%20-%20Spring%20from%20Four%20Seasons.mp3',
        count: 2,
        name: 'Vivaldi - Spring from Four Seasons.mp3',
        ext: 'mp3',
        score: 5 },
      { resource: '/content/README.txt',
        count: 1,
        name: 'README.txt',
        ext: 'txt',
        score: 1 },
      { resource: '/content/throughput-test/10MB.bin',
        count: 1,
        name: '10MB.bin',
        ext: 'bin',
        score: 1 } ],
      topLevelFiles:
      [ { name: 'King James Bible.epub',
        type: 'file',
        mtime: mockDate,
        size: 1457717,
        isTopLevel: false,
        ext: 'epub' },
      { name: 'README.txt',
        type: 'file',
        mtime: mockDate,
        size: 2886,
        isTopLevel: false,
        ext: 'txt' } ] }

    const prepopulatedStore = createStore(reducer, prepopulatedState, compose(applyMiddleware(sagaMiddleware)))
    const component = renderer.create(<Provider store={prepopulatedStore}><Router><App /></Router></Provider>)
    expect(component.toJSON()).toMatchSnapshot()
  })
})
