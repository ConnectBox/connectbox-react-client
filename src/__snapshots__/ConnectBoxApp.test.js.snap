// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectBoxApp tests renders content fully 1`] = `
<div
  style={
    Object {
      "height": "calc(100%)",
    }
  }
>
  <div
    className="content-panel"
  >
    <div
      className="navigation-bar"
    >
      <ol
        className="navbar-container breadcrumb"
        style={
          Object {
            "left": "0px",
            "position": "absolute",
            "right": "0px",
            "zIndex": 1000,
          }
        }
      >
        <li>
          <a
            onClick={[Function]}
            role="button"
            title="Reload"
          >
            <i
              className="fa fa-refresh"
            />
          </a>
        </li>
        <li
          className={null}
          style={Object {}}
        >
          <span>
            <a
              href="/"
              onClick={[Function]}
            >
              <i
                className="fa fa-home fa-lg"
              />
            </a>
          </span>
        </li>
      </ol>
      <div
        className="chat-button"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="fa fa-comments-o fa-lg chat-icon  "
          style={
            Object {
              "display": "none",
              "position": "absolute",
              "right": 10,
              "top": 0,
            }
          }
        />
        <i
          aria-hidden="true"
          className="fa fa-comments fa-lg chat-icon  "
          style={
            Object {
              "display": "none",
              "position": "absolute",
              "right": 10,
              "top": 0,
            }
          }
        />
        <i
          aria-hidden="true"
          className="fa fa-comments fa-lg chat-icon  "
          style={
            Object {
              "display": "block",
              "position": "absolute",
              "right": 10,
              "top": 0,
            }
          }
        />
      </div>
    </div>
    <div
      style={
        Object {
          "marginTop": "3em",
        }
      }
    >
      <div
        className="table-responsive"
      >
        <table
          className="table collapsed"
        >
          <tbody>
            <tr>
              <td
                className="centered"
              >
                <a
                  href="#/Help"
                  onClick={[Function]}
                >
                  <div
                    className="home-icon-container"
                  >
                    <img
                      alt=""
                      src="/content/_icon_Help.png"
                    />
                  </div>
                  <span>
                    Help
                  </span>
                  <span
                    className="directory"
                  />
                </a>
              </td>
            </tr>
            <tr>
              <td
                className="centered"
              >
                <a
                  href="#/Music"
                  onClick={[Function]}
                >
                  <i
                    className="fa fa-folder fa-5x"
                  />
                  <span>
                    <br />
                    Music
                  </span>
                  <span
                    className="directory"
                  />
                </a>
              </td>
            </tr>
            <tr>
              <td
                className="centered"
              >
                <a
                  href="#/Videos"
                  onClick={[Function]}
                >
                  <i
                    className="fa fa-video-camera fa-5x"
                  />
                  <span>
                    <br />
                    Videos
                  </span>
                  <span
                    className="directory"
                  />
                </a>
              </td>
            </tr>
            <tr>
              <td
                className="centered"
              >
                <a
                  href="#/standard-folder"
                  onClick={[Function]}
                >
                  <i
                    className="fa fa-folder fa-5x"
                  />
                  <span>
                    <br />
                    standard-folder
                  </span>
                  <span
                    className="directory"
                  />
                </a>
              </td>
            </tr>
            <tr>
              <td
                className="centered"
              >
                <a
                  href="#/throughput-test"
                  onClick={[Function]}
                >
                  <i
                    className="fa fa-folder fa-5x"
                  />
                  <span>
                    <br />
                    throughput-test
                  </span>
                  <span
                    className="directory"
                  />
                </a>
              </td>
            </tr>
            <tr>
              <td
                className="centered"
              >
                <a
                  href="#/written"
                  onClick={[Function]}
                >
                  <i
                    className="fa fa-book fa-5x"
                  />
                  <span>
                    <br />
                    written
                  </span>
                  <span
                    className="directory"
                  />
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div>
      <div
        className="table-responsive"
      >
        <table
          className="table"
        >
          <tbody>
            <tr>
              <td>
                <a
                  href="/content/King James Bible.epub"
                >
                  <i
                    className="fa fa-file-text-o"
                  />
                  <span
                    className=""
                  >
                    <span>
                       
                      King James Bible.epub
                    </span>
                  </span>
                </a>
              </td>
            </tr>
            <tr>
              <td>
                <a
                  href="/content/README.txt"
                >
                  <i
                    className="fa fa-file-text-o"
                  />
                  <span
                    className=""
                  >
                    <span>
                       
                      README.txt
                    </span>
                  </span>
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div>
      <div
        className="table-responsive"
      >
        <table
          className="table"
        >
          <tbody>
            <tr>
              <td
                className="popular-content-header"
              >
                <i
                  className="fa fa-line-chart"
                />
              </td>
            </tr>
            <tr>
              <td
                style={
                  Object {
                    "borderTop": 0,
                  }
                }
              >
                <a
                  href="/content/Music/Vivaldi%20-%20Spring%20from%20Four%20Seasons.mp3"
                  style={
                    Object {
                      "display": "inline-block",
                    }
                  }
                >
                  <i
                    className="fa fa-file-sound-o"
                  />
                  <span>
                     
                    Vivaldi - Spring from Four Seasons.mp3
                  </span>
                </a>
                <span
                  style={
                    Object {
                      "display": "inline-block",
                      "paddingLeft": "4px",
                    }
                  }
                >
                   
                  <span>
                    <i
                      className="fa fa-flash"
                      style={
                        Object {
                          "color": "#337ab7",
                          "paddingRight": "2px",
                        }
                      }
                    />
                  </span>
                  <span>
                    <i
                      className="fa fa-flash"
                      style={
                        Object {
                          "color": "#337ab7",
                          "paddingRight": "2px",
                        }
                      }
                    />
                  </span>
                  <span>
                    <i
                      className="fa fa-flash"
                      style={
                        Object {
                          "color": "#337ab7",
                          "paddingRight": "2px",
                        }
                      }
                    />
                  </span>
                  <span>
                    <i
                      className="fa fa-flash"
                      style={
                        Object {
                          "color": "#337ab7",
                          "paddingRight": "2px",
                        }
                      }
                    />
                  </span>
                  <span>
                    <i
                      className="fa fa-flash"
                      style={
                        Object {
                          "color": "#337ab7",
                          "paddingRight": "2px",
                        }
                      }
                    />
                  </span>
                </span>
              </td>
            </tr>
            <tr>
              <td
                style={Object {}}
              >
                <a
                  href="/content/README.txt"
                  style={
                    Object {
                      "display": "inline-block",
                    }
                  }
                >
                  <i
                    className="fa fa-file-text-o"
                  />
                  <span>
                     
                    README.txt
                  </span>
                </a>
                <span
                  style={
                    Object {
                      "display": "inline-block",
                      "paddingLeft": "4px",
                    }
                  }
                >
                   
                  <span>
                    <i
                      className="fa fa-flash"
                      style={
                        Object {
                          "color": "#337ab7",
                          "paddingRight": "2px",
                        }
                      }
                    />
                  </span>
                </span>
              </td>
            </tr>
            <tr>
              <td
                style={Object {}}
              >
                <a
                  href="/content/throughput-test/10MB.bin"
                  style={
                    Object {
                      "display": "inline-block",
                    }
                  }
                >
                  <i
                    className="fa fa-file-text-o"
                  />
                  <span>
                     
                    10MB.bin
                  </span>
                </a>
                <span
                  style={
                    Object {
                      "display": "inline-block",
                      "paddingLeft": "4px",
                    }
                  }
                >
                   
                  <span>
                    <i
                      className="fa fa-flash"
                      style={
                        Object {
                          "color": "#337ab7",
                          "paddingRight": "2px",
                        }
                      }
                    />
                  </span>
                </span>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      className="spacer"
    />
    <div
      className="footer breadcrumb"
    >
      <div
        className="spacer"
      />
      <a
        className="settings-button"
        href="/admin/"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="fa fa-cog fa-lg"
        />
      </a>
    </div>
  </div>
</div>
`;
